design/splitfrr-proposal.md

```diff
@@ -0,0 +1,465 @@
+
+- Extend the set of ips we want to advertise to a given neighbor
+Adding an extra neighbor with its set of IPs
+- Extend the set of ips we want to associate a given community to
+- Allow incoming routes for a given neighbor
+
+### Examples
+
+#### Configuration generated by MetalLB
+
+Each MetalLB speaker will generate the configuration related to the node its running on.
+
+For example:
+
+```yaml
+apiVersion: frrk8s.metallb.io/v1alpha1
+kind: FRRConfiguration
+metadata:
+  name: metallb-kind-worker
+  namespace: metallb-system
+spec:
+  nodeSelector:
+    kubernetes.io/hostname: kind-worker
+  bgp:
+    routers:
+    - asn: 64512
+      prefixesIpV4:
+      - 192.168.10.0/32
+      neighbors:
+      - address: 172.18.0.5
+        asn: 64512
+        port: 179
+      - address: 172.18.0.6
+        asn: 64512
+        port: 179
+        toAdvertise:
+          allowed:
+          - 192.168.10.0/32
+```
+
+#### Merging two configurations together
+
+Given the constraint of having the same asn for a given FRR instance (per VRF), every user must agree on
+the ASN given to the FRR instance. Because of this, conflicting configurations will be marked as invalid.
+
+Examples of invalid configurations are:
+
+- Different ASNs for the routers related to the same VRF
+- Different Session Parameters for the router on the same address / port destination
+
+Under this premise, a configuration that allows incoming routes from a given peer would look like:
+
+```yaml
+apiVersion: frrk8s.metallb.io/v1alpha1
+kind: FRRConfiguration
+metadata:
+  name: accept-all-router1
+  namespace: metallb-system
+spec:
+  bgp:
+    routers:
+    - asn: 64512
+      neighbors:
+      - address: 172.18.0.6
+        asn: 64512
+        port: 179
+        toReceive:
+          allowed:
+            allowAll: true
+```
+
+At the same time, another user / controller might decide to advertise another IP to the same neighbor:
+
+```yaml
+apiVersion: frrk8s.metallb.io/v1alpha1
+kind: FRRConfiguration
+metadata:
+  name: advertise-extra-router1
+  namespace: metallb-system
+spec:
+  bgp:
+    routers:
+    - asn: 64512
+      prefixesIpV4:
+      - 192.168.11.0/32
+      neighbors:
+      - address: 172.18.0.6
+        asn: 64512
+        port: 179
+        toAdvertise:
+          allowed:
+          - 192.168.11.0/32
+```
+
+The resulting configuration will be something like:
+
+```yaml
+  bgp:
+    routers:
+    - asn: 64512
+      prefixesIpV4:
+      - 192.168.10.0/32
+      - 192.168.11.0/32
+      neighbors:
+      - address: 172.18.0.5
+        asn: 64512
+        port: 179
+      - address: 172.18.0.6
+        asn: 64512
+        port: 179
+        toAdvertise:
+          allowed:
+          - 192.168.10.0/32
+          - 192.168.11.0/32
+        toReceive:
+          allowed:
+            allowAll: true
+```
+
+### Exposing the status
+
+We should be able to expose:
+
+- The rendered frr configuration for each node
```

Comment:
I'd say this is not strictly necessary (it's more of a debugging thing I feel), maybe nice to have?

Reply:
yep but shouldn't be hard to implement and will be useful